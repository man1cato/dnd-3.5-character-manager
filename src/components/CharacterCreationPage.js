import React from 'react'
import {connect} from 'react-redux'
import Header from './Header'
import {Formik, Form} from 'formik'
import * as Yup from 'yup'

import Page1 from './CreatorFormPage1'
import Page2 from './CreatorFormPage2'
import Page3 from './CreatorFormPage3'
import CreatorFormFooter from './CreatorFormFooter'
import {history} from '../routers/AppRouter'
import {startCreateProfile} from '../actions/profile'
import {getSpecialAbilityIdsFromLevels} from '../utils/getSpecialAbilities'


const pages = [Page1, Page2, Page3]

const validationSchema = Yup.object().shape({
	page1: Yup.object().shape({
		name: Yup.string().min(2, 'Name is too short!').max(30, 'Name is too long!').required('Required'),
		age: Yup.number().required().positive().integer(),
		heightFt: Yup.number('Enter a positive integer').positive('Enter a positive integer').integer('Enter a positive integer').required('Required'),
		heightIn: Yup.number('Enter an integer between 0 and 12').positive('Enter an integer between 0 and 12').integer('Enter an integer between 0 and 12').max(11, 'Enter an integer between 0 and 12').required('Required'),
		weight: Yup.number('Enter a positive integer').positive('Enter a positive integer').integer('Enter a positive integer').required('Required')
	}),
	page2: undefined,
	page3: undefined
})

export class CharacterCreationPage extends React.Component {
	state = {
		selectedRace: this.props.races.find((race) => race.name === 'Human'),
		selectedJobClass: this.props.jobClasses.find((jobClass) => jobClass.name === 'Fighter'),
		page: 1
	}

	handleSelect = (e, setFieldValue, setFieldError) => {
		const name = e.target.name;
		const value = e.target.value;
		if (name === 'race') { setFieldValue('bonusLanguages', []) } 
		if (name === 'school') {
			setFieldValue('prohibitedSchools', []);
			if (value === 'Universal') {
				setFieldError('prohibitedSchools', undefined);
			} 
		} 
		if (name === 'jobClass') {
			if (value === this.props.jobClasses.find((jobClass) => jobClass.name === 'Paladin').id) { 
				setFieldValue('alignment', 'Lawful Good') 
			}
			if (value === this.props.jobClasses.find((jobClass) => jobClass.name === 'Wizard').id) { 
				setFieldValue('school', 'Universal') 
			} else {
				setFieldValue('school', null)
			}
		}
		
		this.setState((prevState) => {
			const selectedRace = name === 'race' ? this.props.races.find((race) => race.id === value) : prevState.selectedRace;
			const selectedJobClass = name === 'jobClass' ? this.props.jobClasses.find((jobClass) => jobClass.id === value) : prevState.selectedJobClass;
			
			return {
				selectedRace,
				selectedJobClass
			}
		});
	}

	handleMultiSelect = (e, setFieldValue) => {
		const name = e.target.name;
		const options = e.target.options;
		const value = [];
		for (let i = 0; i < options.length; i++) {
			if (options[i].selected) {
				value.push(options[i].value);
			}
		}
		setFieldValue(name, value);
	}

	handleBack = (setErrors) => {
		this.setState((prevState) => ({
			page: prevState.page - 1
		}));        
		setErrors({});
	}

	handleNext = () => {
		this.setState((prevState) => ({
			page: prevState.page + 1
		}));
	}    

	render() {        
		return (
			<div className="layout">
				<Header pageTitle="Character Creation" />
				
				<Formik
					initialValues={{
						name: '',
						gender: 'Male',
						age: '',
						heightFt: '',
						heightIn: '',
						weight: '',
						race: this.props.races.find((race) => race.name === 'Human').id,
						jobClass: this.props.jobClasses.find((jobClass) => jobClass.name === 'Fighter').id,
						alignment: 'Lawful Good',
						deity: ''
					}}
												
					validationSchema={Yup.reach(validationSchema, `page${this.state.page}`)}

					onSubmit={(values, {setErrors, setSubmitting}) => {
						getSpecialAbilityIdsFromLevels(values.jobClass, 1).then((specialAbilities) => {
							const profile = {
								...values,
								race: this.state.selectedRace.name,
								size: this.state.selectedRace.size,
								specialAbilities,
								jobClass: this.state.selectedJobClass.name,
								height: `${values.heightFt}'${values.heightIn}"`,
								languages: this.state.selectedRace.defaultLanguages.concat(values.bonusLanguages),
								deity: !!values.deity ? values.deity : "None",
								level: 1,
								iconUrl: this.state.selectedRace.iconUrl
							}
							delete profile.heightFt
							delete profile.heightIn
							if(profile.school === null) {delete profile.school}
							this.props.startCreateProfile(profile)
						})							
						setTimeout(() => { history.push('/profile') }, 2000)
						setSubmitting(false)
					}}
				>

					{({values, setFieldValue, handleChange, isSubmitting, isValid, setErrors, setFieldError}) => (
						<Form >
							<div className="container container--body">
								{{
									1: <Page1 
										values={values}
										races={this.props.races}
										jobClasses={this.props.jobClasses}
										selectedRace={this.state.selectedRace} 
										handleChange={handleChange} 
										handleSelect={this.handleSelect}  
										setFieldValue={setFieldValue}
									/>,
									2: <Page2
										values={values}
										selectedRace={this.state.selectedRace} 
										jobClasses={this.props.jobClasses}
										selectedJobClass={this.state.selectedJobClass}
										alignments={this.props.alignments}
										schools={this.props.schools}
										handleChange={handleChange}
										handleSelect={this.handleSelect}
										handleMultiSelect={this.handleMultiSelect}
										setFieldValue={setFieldValue}
										setFieldError={setFieldError}
									/>,
									3: <Page3
										values={values}
										feats={this.props.feats}
										selectedJobClass={this.state.selectedJobClass}										
										handleChange={handleChange}
										handleSelect={this.handleSelect}
										handleMultiSelect={this.handleMultiSelect}
										setFieldValue={setFieldValue}
										setFieldError={setFieldError}
									/>
								}[this.state.page]}
							</div>
							
							<CreatorFormFooter 
								page={this.state.page}
								pages={pages}
								handleBack={this.handleBack}
								handleNext={this.handleNext}
								handleSubmit={this.handleSubmit}
								setErrors={setErrors}
								isSubmitting={isSubmitting}
								isValid={isValid}
							/>

						</Form>
					)}
				</Formik>
					
			</div>
		)
	}
};


const mapStateToProps = (state) => ({
	races: state.races,
	jobClasses: state.jobClasses,
	feats: state.feats,
	schools: [
		'Abjuration','Clairsentience', 'Conjuration', 'Divination', 
		'Enchantment', 'Evocation', 'Illusion', 'Metacreativity', 
		'Necromancy', 'Psychokinesis', 'Psychometabolism', 'Psychoportation', 
		'Telepathy', 'Transmutation', 'Universal'
	],
	alignments: [
		'Lawful Good', 'Neutral Good', 'Chaotic Good', 
		'Lawful Neutral', 'True Neutral', 'Chaotic Neutral', 
		'Lawful Evil', 'Neutral Evil', 'Chaotic Evil'
	]
});

const mapDispatchToProps = (dispatch, props) => ({
	startCreateProfile: (profile) => dispatch(startCreateProfile(profile))
});


export default connect(mapStateToProps, mapDispatchToProps)(CharacterCreationPage);
