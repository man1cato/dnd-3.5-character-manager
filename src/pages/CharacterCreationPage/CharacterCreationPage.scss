@import '../../styles/base/settings';


.CreatorForm {
   height: 100vh;
   display: flex;
   flex-direction: column;
   overflow: hidden;
}

.CreatorForm__group {
   display: flex;
   justify-content: space-between;
   align-items: center;
   margin-bottom: $m-size;
}

.CreatorForm__input-group {
   display: flex;
   align-items: center;
   > * {
      margin-right: $s-size;
   }
   input {
      max-width: 70px; 
   }
   > :last-child {
      margin-right: 0;
   }
}

.CreatorForm__grid {
	display: grid;
   grid-template-rows: 2rem auto;
	grid-gap: $xs-size;
	justify-items: center;
	align-items: center;
	text-align: center; 
   font-size: $font-size-m;
}

.CreatorForm__error {
   font-style: italic;
   font-size: $font-size-s;
   margin: 0;
}

.CreatorForm__error--center {
   @extend .CreatorForm__error;
   text-align: center;
}

.CreatorForm__error--right {
   @extend .CreatorForm__error;
   text-align: right;
}