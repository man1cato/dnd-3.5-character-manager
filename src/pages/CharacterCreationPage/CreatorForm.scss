@import '../../styles/base/settings';


.CreatorForm {
   height: 100vh;
   display: flex;
   flex-direction: column;
   overflow: hidden;
}

.CreatorForm__scroll-container {
   flex-grow: 1;
   display: flex;
   flex-direction: column;
   overflow: hidden;
}

.CreatorForm__scroll-list {
   flex-grow: 1;
   overflow-y: auto;
   padding-top: $xs-size;
}

.CreatorForm__grid {
	display: grid;
   grid-template-rows: 2rem auto;
	grid-gap: $xs-size;
	justify-items: center;
	align-items: center;
	text-align: center; 
   font-size: $font-size-m;
}

.CreatorFormSkills {
   display: flex;
   flex-direction: column;
   overflow: hidden;
}

.CreatorFormSkills__grid {
   @extend .CreatorForm__grid;
   grid-template-columns: 2fr 1fr;
   grid-template-rows: auto;
   > button {
      padding-bottom: 0;
   }
}
.CreatorFormSkills__grid--header {
   @extend .CreatorFormSkills__grid;
   padding-right: 17px;
   padding-bottom: 0.5rem;
   border-bottom: 1px solid $dark-grey;
}

.CreatorFormSkills__counter {
   height: 4rem;
   button {
      height: 100%;
      padding: $xs-size !important; 
   }
   input {
      height: 100%;
      width: 100%;
	}
}


.CreatorFormEquipment__grid {
   @extend .CreatorForm__grid;
   grid-template-columns: 4fr repeat(3, 2fr) 1fr;
   input {
      max-width: 3rem;
   }
}


.CreatorFormFeats__list {
   display: flex;
   justify-content: space-between;
   align-items: center;
   margin-bottom: $m-size;
   
   > :first-child {
      flex: 1 1 80%;
      padding: 0 $s-size 0 0;
   }
   > :last-child {
      flex: 1 0 100px;   
   }
}


.CreatorForm__error {
   font-style: italic;
   font-size: $font-size-s;
   margin: 0;
}

.CreatorForm__error--right {
   @extend .CreatorForm__error;
   text-align: right;
}


.CreatorFormFooter {
   position: fixed;
   bottom: 0px;
   height: 10vh;
   min-height: 60px;
   width: -webkit-fill-available;
   background: white;
   box-shadow: 0 0 5px $light-grey;
   padding: $s-size 0;
}

.CreatorFormFooter__content {
   display: flex;
   justify-content: space-between;
   > * {
      width: -webkit-fill-available;
   }
   :nth-child(2) {
      margin-left: $s-size;
   }
}