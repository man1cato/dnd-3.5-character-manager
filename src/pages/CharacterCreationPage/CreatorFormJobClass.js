import React, { useEffect } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'

import { schools, alignments } from '../../utils/staticData'
import { apiObjectToArray, calcStartingGold } from '../../utils/utils'


const validateProhibitedSchools = (selectedSchool) => (value) => {
	let error
	if (selectedSchool === 'Divination') {
		if (value.length < 1) {
			error = 'You must select one school whose magic use will be prohibited'
		}
		if (value.length > 1) {
			error = 'Select only one school whose magic use will be prohibited'
		}
	} else {
		if (selectedSchool !== 'Universal' && value.length < 2) {
			error = 'You must select two schools whose magic use will be prohibited'
		}
		if (selectedSchool !== 'Universal' && value.length > 2) {
			error = 'Select only two schools whose magic use will be prohibited'
		}
	}
	return error
}

const CreatorFormJobClass = ({
	values, 
	selectedRace,
	selectedJobClass, 
	setSelectedJobClass,
	jobClasses, 
	handleChange, 
	handleMultiSelect, 
	setFieldValue, 
	setFieldError
}) => {
	const { school } = values

	useEffect(() => {
		if (selectedJobClass.name === 'Paladin') {
			setFieldValue('alignment', 'Lawful Good')
		} else if (selectedJobClass.name === 'Wizard') {
			setFieldValue('school', 'Universal')
		} else {
			setFieldValue('school', null)
			setFieldValue('prohibitedSchools', [])
		}
	}, [selectedJobClass])

	return (
		<div className="container--body">			
			<div className="form-group--35">
				<h4>Job Class:</h4>
				<Field
					className="select"
					name="jobClass"
					component="select"
					onChange={(e) => {
						setSelectedJobClass(jobClasses[e.target.value])
						setFieldValue('startingGold', 0)
						handleChange(e)
					}}
				>
					{apiObjectToArray(jobClasses).map((jobClass, i) => (
						<option
							key={`jobClass${i}`}
							value={jobClass.id}
						>
							{jobClass.name}
						</option>
					))}
				</Field>
			</div>

			<div className="form-group--35">
				<h4>Alignment:</h4>
				{selectedJobClass.name === "Paladin" ?
					<div>Lawful Good</div>
					:
					<Field 
						className="select" 
						name="alignment" 
						component="select" 
					>
						{alignments.map((alignment, i) => (
							<option 
								value={alignment} 
								key={`alignment${i}`}
							>
								{alignment}
							</option>
						))}
					</Field>
				}
			</div>		

			<div className="form-group--35 align-top">
				<h4>Hit Die:</h4>
				<div>{selectedJobClass.hitDie}</div>
			</div>

			<div className="form-group--35">
				<h4>Starting Gold:</h4>
				<div className="input-group">
					<input
						className="number-input"
						type="number"
						value={values.startingGold}
						onChange={(e) => setFieldValue('startingGold', e.target.value)}
					/>
					<div> gp</div>
					<button
						className="button"
						type="button"
						onClick={() => setFieldValue('startingGold', calcStartingGold(selectedJobClass.name))}
					>
						Roll
					</button>
				</div>
			</div>

			<div className="form-group--35 align-top">
				<h4>Proficiencies:</h4>
				<div>{selectedJobClass.proficiencies.join(", ")}</div>
			</div>

			<div className="form-group--35 align-top">
				<h4>Bonus Languages:</h4>
				<Field 
					className="multi-select" 
					name="bonusLanguages" 
					component="select" 
					multiple 
					onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
				>
					{_.orderBy(selectedRace.bonusLanguages).map((language) => (
						<option 
							value={language} 
							key={language}
						>
							{language}
						</option>
					))}
				</Field>                                    
			</div>

			<div className="form-group--35">
				<h4>Deity:</h4>
				<Field 
					className="text-input" 
					name="deity" 
					placeholder="(Optional) Enter the name of a Deity" 
				/>                                  
			</div>
		
			{selectedJobClass.name === 'Wizard' && (
				<div className="form-group--35">
					<h4>School:</h4>
					<Field 
						className="select" 
						name="school" 
						component="select" 
						onChange={(e) => {
							if (e.target.value === 'Universal') { 
								setFieldError('prohibitedSchools', undefined)	
								setFieldValue('prohibitedSchools', [])
							}						
							handleChange(e)
						}}
					>
						{schools.map((school, i) => (
							<option 
								value={school} 
								key={`school${i}`}
							>
								{school}
							</option>
						))}
					</Field>
				</div>
			)}

			{selectedJobClass.name === 'Wizard' && school !== 'Universal' && (
				<>
					<div className="form-group--35 align-top">
						<h4>Prohibited School(s):</h4>
						<Field 
							className="multi-select"
							name="prohibitedSchools" 
							component="select" 
							multiple 
							validate={validateProhibitedSchools(school)}
							onChange={(e) => {handleMultiSelect(e, setFieldValue)}}
						>
							{_.difference(schools, [school, 'Universal', 'Divination']).map((school, i) => (
								<option 
									value={school} 
									key={`prohibited${i}`}
								>
									{school}
								</option>
							))}
						</Field>
					</div>
					<ErrorMessage className="form-group--error" name="prohibitedSchools" component="div"/>
				</>
			)}
		</div>
	)
}

export default CreatorFormJobClass
