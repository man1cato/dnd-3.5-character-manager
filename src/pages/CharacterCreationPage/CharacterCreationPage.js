import React, { useState } from 'react'
import { connect } from 'react-redux'
import { Formik, Form } from 'formik'
import * as Yup from 'yup'
import _ from 'lodash'

import { history } from '../../routers/AppRouter'
import { startCreateProfile } from '../../store/actions/profile'
import { apiObjectToArray, calcAbilityMod, convertMoneyToDenominations, calcAttackBonuses } from '../../utils/utils'
import { abilities, sizeMods } from '../../utils/staticData'

import CreatorFormIdentity from './CreatorFormIdentity/CreatorFormIdentity'
import CreatorFormJobClass from './CreatorFormJobClass/CreatorFormJobClass'
import CreatorFormAbilities from './CreatorFormAbilities/CreatorFormAbilities'
import CreatorFormFeats from './CreatorFormFeats/CreatorFormFeats'
import CreatorFormSkills from './CreatorFormSkills/CreatorFormSkills'
import CreatorFormEquipment from './CreatorFormEquipment/CreatorFormEquipment'
import CreatorFormFooter from './CreatorFormFooter/CreatorFormFooter'


const pages = [CreatorFormIdentity, CreatorFormJobClass, CreatorFormAbilities, CreatorFormFeats, CreatorFormSkills, CreatorFormEquipment]

const abilityMessage = 'Enter an integer between 3 and 18'
const abilityValidation = Yup.number().integer(abilityMessage).min(3, abilityMessage).max(18, abilityMessage).required('Required').typeError(abilityMessage)
const validationSchema = Yup.object().shape({
	page1: Yup.object().shape({
		name: Yup.string().max(30, 'Name is too long!').required('Required').typeError('Required'),
		age: Yup.number().positive().integer().required('Required').typeError('Enter a number'),
		height: Yup.object().shape({
			ft: Yup.number().positive('Enter a positive integer').integer('Enter a positive integer').required('Required').typeError('Enter a number').typeError('Enter a number'),
			in: Yup.number().positive('Enter an integer between 0 and 12').integer('Enter an integer between 0 and 12').max(11, 'Enter an integer between 0 and 12').required('Required').typeError('Enter a number'),
		}),
		weight: Yup.number().positive('Enter a positive integer').integer('Enter a positive integer').required('Required').typeError('Enter a number')
	}),
	page2: Yup.object().shape({
		startingGold: Yup.number().positive('Cannot be negative').integer('Enter an integer')
	}),
	page3: Yup.object().shape({
		abilities: Yup.object().shape(
			_.mapValues(abilities, () => Yup.object().shape({ score: abilityValidation }))
		)
	}),
	page4: Yup.object().shape({
		feats: Yup.array().required('Select at least one feat')
	}),
	page5: Yup.object().shape({
		remainingSkillPoints: Yup.number().moreThan(-1, 'Too many skill points assigned').lessThan(1, 'Assign all skill points')
	}),
	page6: Yup.object().shape({
		equipment: Yup.array().required('Select at least one equipment'),
		remainingGold: Yup.number().moreThan(-1, 'Purchases must not exceed available money')
	})
})


export const CharacterCreationPage = props => {
	const [page, setPage] = useState(1)

	const handleMultiSelect = (e, setFieldValue) => {
		const name = e.target.name
		const options = e.target.options
		const value = []
		for (let i = 0; i < options.length; i++) {
			if (options[i].selected) {
				value.push(options[i].value)
			}
		}
		setFieldValue(name, value)
	}
	
	return (
		<Formik
			initialValues={{
				name: '',
				gender: 'Male',
				age: '',
				height: {
					ft: '',
					in: ''
				},
				weight: '',
				alignment: 'Lawful Good',
				race: _.find(props.races, race => race.name === 'Human'),
				jobClass: _.find(props.jobClasses, jobClass => jobClass.name === 'Fighter'),
				bonusLanguages: [],
				deity: '',
				abilities: _.mapValues(abilities, () => ({ score: '', final: '' })),
				feats: [],
				equipment: [],
				equipped: {
					armor: null,
					shield: null,
					weapons: []
				},
				skillSet: apiObjectToArray(props.skills).map(skill => ({id: skill.id, ranks: 0 })),
				skillPoints: 0,
				remainingSkillPoints: 0,
				startingGold: 0,
				remainingGold: 0
			}}
										
			validationSchema={Yup.reach(validationSchema, `page${page}`)}

			onSubmit={(values, { setSubmitting }) => {
				const { 
					name, gender, age, height, weight, race, alignment, jobClass, 
					remainingGold, bonusLanguages, deity, feats, skillSet,
					equipment, equipped, school, prohibitedSchools  
				} = values

				const abilities = _.mapValues(values.abilities, ability => ({ 
					score: ability.final,
					tempScore: ''
				}))
				const dexMod = calcAbilityMod(abilities.dex.score)
				let armorBonus = 0
				if (equipped.armor) { armorBonus += props.items[equipped.armor].armorBonus }
				if (equipped.shield) { armorBonus += props.items[equipped.shield].armorBonus }
				const baseArmorClass = 10 + armorBonus + sizeMods[race.size].size + dexMod
				
				const profile = {
					name,
					gender,
					age,
					height: `${height.ft}'${height.in}"`,
					weight,
					race: _.findKey(props.races, race),
					alignment,
					jobClass: _.findKey(props.jobClasses, jobClass),
					money: convertMoneyToDenominations(remainingGold),
					languages: _.sortBy(race.defaultLanguages.concat(bonusLanguages)),
					specialAbilities: jobClass.levels["1"].specialAbilities,
					deity: !!deity ? deity : "None",
					abilities,
					feats,
					skillSet: _.filter(skillSet, skill => skill.ranks > 0),
					equipment,
					equipped,
					ac: {
						base: baseArmorClass,
						flat: baseArmorClass - dexMod,
						touch: baseArmorClass - armorBonus
					},
					attackBonuses: _.mapValues(
						calcAttackBonuses(
							jobClass.levels[1].baseAttackBonuses || [0],
							race.size,
							abilities.str.score,
							abilities.dex.score
						),
						val => ({ base: val, mod: 0, total: val })
					),
					level: 1,
					hp: {
						base: Number(jobClass.hitDie.slice(1)) + calcAbilityMod(abilities.con.score)
					},
					xp: 0,
					iconUrl: race.iconUrl
				}
				if(!!school) {
					profile.school = school 
					profile.prohibitedSchools = prohibitedSchools
				}
				console.log('Created profile: ', profile)
				props.startCreateProfile(profile)
				
				setTimeout(() => { history.push('/profile') }, 1500)
				setSubmitting(true)
			}}
		>
			{({ values, setFieldValue, handleChange, handleSubmit, isSubmitting, isValid, validateForm, setErrors, setFieldError, setTouched}) => (
				<Form className="CreatorForm">
					{{
						1: <CreatorFormIdentity 
							values={values}
							races={props.races}
							jobClasses={props.jobClasses}
							setFieldValue={setFieldValue}
							setTouched={setTouched}
						/>,
						2: <CreatorFormJobClass
							values={values}
							jobClasses={props.jobClasses}
							handleChange={handleChange}
							handleMultiSelect={handleMultiSelect}
							setFieldValue={setFieldValue}
							setFieldError={setFieldError}
							validateForm={validateForm}
						/>,
						3: <CreatorFormAbilities
							values={values}
							handleChange={handleChange}
							setFieldValue={setFieldValue}
							validateForm={validateForm}								
						/>,
						4: <CreatorFormFeats
							values={values}
							feats={props.feats}
							setFieldValue={setFieldValue}
						/>,
						5: <CreatorFormSkills
							values={values}
							skills={props.skills}
							setFieldValue={setFieldValue}
						/>,
						6: <CreatorFormEquipment
							values={values}
							items={props.items}
							setFieldValue={setFieldValue}
						/>
					}[page]}

					<CreatorFormFooter 
						page={page}
						pageCount={pages.length}
						setPage={setPage}
						handleSubmit={handleSubmit}
						setErrors={setErrors}
						isSubmitting={isSubmitting}
						isValid={isValid}
						validateForm={validateForm}
					/>

				</Form>
			)}
		</Formik>				
	)
}


const mapStateToProps = state => ({
	races: state.races,
	jobClasses: state.jobClasses,
	feats: _.omitBy(state.feats, feat => feat.prerequisites || _.includes(feat.types, 'Epic') || _.includes(feat.types, 'Creature')),
	items: _.omitBy(state.items, item => item.weaponType === 'Natural' || _.includes(['Creature Part', 'Money'], item.category)),
	skills: state.skills
})

const mapDispatchToProps = dispatch => ({
	startCreateProfile: profile => dispatch(startCreateProfile(profile))
})


export default connect(mapStateToProps, mapDispatchToProps)(CharacterCreationPage)