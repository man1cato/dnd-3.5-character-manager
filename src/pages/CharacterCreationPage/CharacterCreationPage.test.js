import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import _ from 'lodash'

import { apiData, renderWithRedux } from '../../tests/utils'
import { CharacterCreationPage } from './CharacterCreationPage'


const startCreateProfile = jest.fn()

let props

beforeAll(async () => {
	const api = await apiData()
	props = {
		races: api.races,
		jobClasses: api.jobClasses,
		feats: _.omitBy(api.feats, feat => feat.prerequisites || _.includes(feat.types, 'Epic') || _.includes(feat.types, 'Creature')),
		items: _.omitBy(api.items, item => item.weaponType === 'Natural' || _.includes(['Creature Part', 'Money'], item.category)),
		skills: api.skills,
		startCreateProfile
	}
})

test('should render default state of CharacterCreationPage correctly', () => {
	const { container } = render(<CharacterCreationPage {...props} />)
	expect(container.firstChild).toMatchSnapshot()
})

test('should make form page 1 valid after inputting required fields', () => {
	// expect(wrapper.find(CreatorFormFooter).prop('isValid')).toBe(false)

	// wrapper.find('input').at(0).simulate("change", { target: { id: '1', name: 'name', value: "Arc" } })
	// wrapper.find('input').at(1).simulate("change", { target: { id: '2', name: 'age', value: "21" } })
	// wrapper.find('input').at(2).simulate("change", { target: { id: '3', name: 'height.ft', value: "4" } })
	// wrapper.find('input').at(3).simulate("change", { target: { id: '4', name: 'height.in', value: "7" } })
	// wrapper.find('input').at(4).simulate("change", { target: { id: '5', name: 'weight', value: "153" } })

	// expect(wrapper.find(CreatorFormFooter).prop('isValid')).toBe(true)
})

test('should render page 2 when page 1 is valid and next is clicked', () => {
	// expect(wrapper.find(CreatorFormIdentity)).toHaveLength(1)
	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(0)
	// expect(wrapper.find(CreatorFormFooter).prop('isValid')).toBe(true)

	// wrapper.find('#nextButton').simulate('click')

	// expect(wrapper.find(CreatorFormIdentity)).toHaveLength(0)
	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(1)
	// expect(wrapper.find('#submitButton')).toHaveLength(0)
})

test('should go back a page when back button is clicked', () => {
	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(1)

	// wrapper.find('#backButton').simulate('click')

	// expect(wrapper.find(CreatorFormIdentity)).toHaveLength(1)
	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(0)

	// wrapper.find('#nextButton').simulate('click')
	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(1)
})

test('should render page 3 when page 2 is valid and next is clicked', () => {	
	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(1)
	// expect(wrapper.find(CreatorFormFooter).prop('isValid')).toBe(true)
	
	// wrapper.find('#nextButton').simulate('click')

	// expect(wrapper.find(CreatorFormJobClass)).toHaveLength(0)
	// expect(wrapper.find(CreatorFormAbilities)).toHaveLength(1)
	// expect(wrapper.find('#submitButton')).toHaveLength(0)
})
 
// test('should render page 4 when page 3 is valid and next is clicked', () => {
// 	expect(wrapper.find(CreatorFormAbilities)).toHaveLength(1)
// 	expect(wrapper.find(CreatorFormFooter).prop('isValid')).toBe(false)

// 	wrapper.find('#abilitiesRollButton').simulate('click')
	
// 	expect(wrapper.find(CreatorFormFooter).prop('isValid')).toBe(true)
	
// 	wrapper.find('#nextButton').simulate('click')
	
// 	expect(wrapper.find(CreatorFormAbilities)).toHaveLength(0)
// 	expect(wrapper.find(CreatorFormFeats)).toHaveLength(1)
// })

// test('should call startCreateProfile when on final page of form and submit button is clicked', () => {
// 	const pageCount = wrapper.find(CreatorFormFooter).props().pages.length
	
// 	expect(wrapper.find('#nextButton')).toHaveLength(1)
	
// 	wrapper.find(CharacterCreationPage).setState({ page: pageCount })

// 	expect(wrapper.find('#nextButton')).toHaveLength(0)
// 	expect(wrapper.find('#submitButton')).toHaveLength(1)
	
// 	wrapper.find('#submitButton').simulate('click')

// 	expect(wrapper.find(Provider).prop('store').getState()).toHaveProperty('profile')
// })