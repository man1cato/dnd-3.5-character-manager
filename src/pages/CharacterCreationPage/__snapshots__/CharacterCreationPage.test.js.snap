// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render default state of CharacterCreationPage correctly 1`] = `
<form
  class="CreatorForm"
>
  <div
    class="container--body"
  >
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Name:
      </h4>
      <div>
        <input
          class="text-input"
          name="name"
          value=""
        />
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Gender:
      </h4>
      <select
        class="select"
        name="gender"
      >
        <option
          value="Male"
        >
          Male
        </option>
        <option
          value="Female"
        >
          Female
        </option>
        <option
          value="Non-Binary"
        >
          Non-Binary
        </option>
      </select>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Age:
      </h4>
      <div>
        <div
          class="CreatorForm__input-group"
        >
          <input
            class="text-input"
            name="age"
            type="number"
            value=""
          />
          <div>
             yrs
          </div>
        </div>
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Height:
      </h4>
      <div>
        <div
          class="CreatorForm__input-group"
        >
          <input
            class="text-input"
            name="height.ft"
            type="number"
            value=""
          />
          <div>
            ft
          </div>
          <input
            class="text-input"
            name="height.in"
            type="number"
            value=""
          />
          <div>
            in
          </div>
        </div>
        <div
          class="CreatorForm__input-group"
        />
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Weight:
      </h4>
      <div>
        <div
          class="CreatorForm__input-group"
        >
          <input
            class="text-input"
            name="weight"
            type="number"
            value=""
          />
          <div>
            lbs
          </div>
        </div>
      </div>
    </div>
    <div
      class="divider"
    />
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Race:
      </h4>
      <select
        class="select"
      >
        <option
          value="recOadObhQKOMa7f2"
        >
          Gray Elf
        </option>
        <option
          value="recUDu56yAAHN3Ue3"
        >
          Half-Elf
        </option>
        <option
          value="recWfP0nya01jKP44"
        >
          Half-Orc
        </option>
        <option
          value="recw2GdtKxEQhIpGm"
        >
          High Elf
        </option>
        <option
          value="recv3zpW98noaCctp"
        >
          Hill Dwarf
        </option>
        <option
          selected=""
          value="recxyRHegDBrFYJP5"
        >
          Human
        </option>
        <option
          value="recFEpQ9zuD9Yqu7l"
        >
          Lightfoot Halfling
        </option>
        <option
          value="recrsiTIDAQXD6mCo"
        >
          Mountain Dwarf
        </option>
        <option
          value="recoDD3gAaQXU3A56"
        >
          Orc
        </option>
        <option
          value="recd3zkVkYKZJP92G"
        >
          Rock Gnome
        </option>
        <option
          value="recGQmGZazcjLyEXv"
        >
          Wild Elf
        </option>
        <option
          value="recEPV5rASoteAGoq"
        >
          Wood Elf
        </option>
      </select>
    </div>
    <div
      class="CreatorFormIdentity__group align-bottom"
    >
      <h4>
        Favored Class:
      </h4>
      <div>
        Any
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Size:
      </h4>
      <div>
        Medium
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Speed:
      </h4>
      <div>
        30
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--bottom"
    >
      <h4>
        Racial Modifiers:
      </h4>
      <div
        class="CreatorForm__input-group"
      >
        <div>
          None
        </div>
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--bottom"
    >
      <h4>
        Default Language(s):
      </h4>
      <div>
        Common
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--top"
    >
      <h4>
        Bonus Languages:
      </h4>
      <div>
        Draconic, Dwarven, Elven, Giant, Gnoll, Gnome, Goblin, Orc, Halfling, Sylvan, Undercommon, Terran, Abyssal
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--top"
    >
      <h4>
        Racial Bonuses:
      </h4>
      <ul
        class="CreatorFormIdentity__list"
      >
        <li>
          1 extra feat at 1st level.
        </li>
        <li>
          4 extra skill points at 1st level and 1 extra skill point at each additional level. 
        </li>
      </ul>
    </div>
    <div
      class="CreatorFormIdentity__group--top"
    >
      <h4>
        Special Abilities:
      </h4>
      <div>
        None
      </div>
    </div>
  </div>
  <footer
    class="CreatorFormFooter"
  >
    <div
      class="container CreatorFormFooter__content"
    >
      <button
        class="button--form-nav button--primary"
        disabled=""
        id="nextButton"
        type="button"
      >
        Next
      </button>
    </div>
  </footer>
</form>
`;
